x-pack/plugins/security_solution/server/lib/detection_engine/signals/source_fields_merging/README.md

```diff
@@ -0,0 +1,389 @@
+```
+source        | fields        | value after merge
+-----         | ---------     | -----
+undefined     | undefined     | undefined
+undefined     | f_[]          | undefined
+p_[]          | undefined     | p_[]
+p_[]          | f_[]          | p_[]
+p_p1          | undefined     | p_p1
+p_p1          | f_[]          | p_p1
+p_[p1]        | undefined     | p_[p1]
+p_[p1]        | f_[]          | p_[p1]
+p_[p1, ...1]  | undefined     | p_[p1, ...1]
+p_[p1, ...1]  | f_[]          | p_[p1, ...1]
+p_{}1         | undefined     | p_{}1
+p_{}1         | f_[]          | p_{}1
+p_[{}1]       | undefined     | p_{}1
+p_[{}1]       | f_[]          | p_{}1
+p_[{}1, ...1] | undefined     | p_[{}1, ...1]
+p_[{}1, ...1] | f_[]          | p_[{}1, ...1]
+f_[]          | undefined     | f_[]
+f_[]          | f_[]          | f_[]
+f_p1          | undefined     | f_p1
+f_p1          | f_[]          | f_p1
+f_[p1]        | undefined     | f_[p1]
+f_[p1]        | f_[]          | f_[p1]
+f_[p1, ...1]  | undefined     | f_[p1, ...1]
+f_[p1, ...1]  | f_[]          | f_[p1, ...1]
+f_{}1         | undefined     | f_{}1
+f_{}1         | f_[]          | f_{}1
+f_[{}1]       | undefined     | f_{}1
+f_[{}1]       | f_[]          | f_{}1
+f_[{}1, ...1] | undefined     | f_[{}1, ...1]
+f_[{}1, ...1] | f_[]          | f_[{}1, ...1]
+```
+
+When source key and source value does not exist but field keys and values do exist, then you
+you will always get field keys and values replacing the source key and value. Caveat is that
+fields will create a single item rather than an array item if field keys and value only has a single
+array element. Also, we prefer to create an object structure in source (e.x. p_p2 instead of a flattened object f_p2)
+for the `merge_all_fields_with_source` strategy
+```
+source        | fields        | value after merge
+-----         | ---------     | -----
+undefined     | f_[p2]        | p_p2         <-- Unboxed from array
+undefined     | f_[p2, ...2]  | p_[p2, ...2]
+undefined     | f_[{}2]       | p_{}2        <-- Unboxed from array
+undefined     | f_[{}2, ...2] | p_[{}2, ...2]
+```
+
+For the `merge_missing_fields_with_source` it will be that we completely skip the fields that contain nested
+fields or type fields such as geo points.
+
+```
+source     | fields        | value after merge
+-----      | ---------     | -----
+undefined  | f_[p2]        | p_p2         <-- Unboxed from array
+undefined  | f_[p2, ...2]  | p_[p2, ...2]
+undefined  | f_[{}2]       | {}           <-- We have an empty object since we only merge primitives
+undefined  | f_[{}2, ...2] | {}           <-- We have an empty object since we only merge primitives
+```
+
+When source key is either a primitive key or a flattened object key with a primitive value (p_p1 or f_p1),
+then we overwrite source value with fields value as an unboxed value array if fields value is a
+single array element (f_[p2] or f[{}2]), otherwise we overwrite source as an array.
```

Comment:
As above, can you group these by algorithm? This is a _fantastic_ reference doc, but grouping by capability here will go a long way with folks groking it all 😅

Reply:
Yes, I will update the docs here as well.